### ISSUE TITLE
Auto fetch updates

### ISSUE THREAD

**Issue by JojiiOfficial** [anatol/pacoloco#14]
What about a feature, which makes the proxy more of an "intelligent" proxy. The proxy detects the downloaded package. If a new version of this package exists, the proxy downloads this new version, even if it hasn't been requested by a client. If this new version won't be used/downloaded, upcoming versions won't be predownloaded. This improves updates even at the first download.

**Comment by marmistrz** [comment]
What about large packages that have been downloaded, removed locally and never to be downloaded again? 
But I agree that it would be useful to fetch the updates overnight and only download them from pacoloco during the day.

**Comment by JojiiOfficial** [comment]
> What about large packages that have been downloaded, removed locally and never to be downloaded again?

What exactly do you mean by that?

**Comment by marmistrz** [comment]
Suppose you download a 350M game and decide you don't like it, uninstall it. And the game is rebuilt every week. This game will still be downloaded by pacoloco and will generate a lot of unnecessary transfer, because it'll never be downloaded again.

**Comment by JojiiOfficial** [comment]
An idea would be to use configurations for each repo. This way you can set different 'max upgrade sizes' individually for your repos. One more configuration could be to set a threshold for how many further upgrades should be pre-cached by pacoloco without downloading. That would allow to pre-cache eg. up to 3 further versions of core + extra packages and max 1 further version of community packages.

**Comment by marmistrz** [comment]
yes, that's what I mean. max upgrade size it tricky: if a package is frequently rebuilt, it will cause substantial traffic anyway. But a package whitelist would probably a good approach for the PoC.

**Comment by JojiiOfficial** [comment]
With max upgrade-size I mean a limit for individual package-sizes which will be pre-cached. For example if downloading 1GB takes up to 1h in your network you can say you want to precache packages only up to 250M to prevent big updates from taking down your whole internet connection.
I think this should be implemented, since this will prevent accidental big downloads and whole takedowns of your internet.
In addition maybe a time based bandwidth-limit option would be a nice feature. This way you can say at night the packages can download at full speed and during the day it should download at 1MB/s.

**Comment by marmistrz** [comment]
I definitely want to precache large packages such as the kernels, if I'm sure I'll downloaded them anyway.

**Comment by JojiiOfficial** [comment]
Me too. This is why I would set the default value to ~2G-2.5G

**Comment by Bradan** [comment]
If there was no request incoming at pacoloco for a specific package for a specific amount of time, it should be deleted anyway from the cache. This would solve the large package problem.

**Comment by JojiiOfficial** [comment]
@Bradan 
Imo using a "max cache limit" would be helpful. If more space is required than specified, old packages will be removed.

**Comment by Bradan** [comment]
> @Bradan
> Imo using a "max cache limit" would be helpful. If more space is required than specified, old packages will be removed.

I haven't excluded your "max cache limit". However it is no solution to the problem above mentioned by @marmistrz, is it?
You can still have the "max cache limit" (I would also need it) but this would be a different feature request, not related to solving this problem.

**Comment by TheAlgorythm** [comment]
Is this active?
I haven't used pacoloco but I am interested in this particular feature.
A beta without any restrictions could be a good start?

**Comment by anatol** [comment]
AFAIK nobody works on this issue.

Feel free to pick it up if you are interested to implement this feature.

**Comment by Focshole** [comment]
I am working on it, I hope to finish it by today (some tests are missing). It is not a super clean work (downloadFile used to do many things, so I had to copy it with a single writer), but it should work.

**Comment by anatol** [comment]
The general idea sounds fine to me.

One question I have is "how do you determine which files needs to be prefetched?"

**Comment by Focshole** [comment]
The logic I have thought is about setting a TTL on each package. The optimal one may be by setting a hit counter, and each time keep the ones with the highest number of hits (e.g. 90% of them). Each time it gets downloaded, i'll reset it to the default value (at the moment it is 5). The TTL value gets decremented before each prefetch. So far, the prefetch can happen weekly, monthly or daily. I may add an option like "every n days".
 
The optimal solution may be by setting a hit counter, and each time keep the ones with the highest number of hits (e.g. before each prefetch, keep 99% of them).

**Comment by Focshole** [comment]
What I haven't considered so far is that the url changes per each version. So far I have been writing it considering to keep the url as index, but it cannot work, so I have to find a way to know the newer version upstream url per each package.

**Comment by anatol** [comment]
Yeah, this 'determine the next file url' is the trickiest part of this issue. I see 2 ways to implement it:

1. for a current version `x.y.z` check if `x.y.z+1` exist at the server. There are multiple corner cases here e.g. what is published version bumped major version, or version jumped by 2 (i.e. `x.y.z+2`) or `pkgrel`/`epoch` changed rather than the version.
2. On database download you need to unpack/parse it and then detect any version bumps for packages pacoloco is interested in. btw the pacman database format is quite simple and you can implement it easily with go, I do not think you really need `go-alpm` here.

**Comment by Focshole** [comment]
Probably 1. is not feasible (i saw sometimes a weird numbering such as `x.y-z` or `1.2.003`) but 2. should work. I'll work on it in my free time

**Comment by Focshole** [comment]
Ok, good news, I have finished writing the code and it seems to work. Some tests are missing and are poorly written, but they should work. There are no breaking changes with the previous version, with a previous version's config file, it would run exactly as before.
 
Cannot guarantee the prefetching's reliability, tests have many TODOs around, but I don't expect it to fail too much.

I am testing it at home right now with default repositories, I'll provide a few updates in the following days (I plan to prefetch every night for a week to check if it does a good work or not).

Till now I have found maybe a corrupted file, which could have been caused by a stopped download but I'm not sure about it (so, it shouldn't be prefetching related). 

I have added a sqlite3 database to keep track of the downloaded databases links, integrated with GORM.  Couldn't keep all the relevant data in memory cause it could have been a bit overkill as memory consumption and even harder to handle. This could be useful for deploying the "stats" thing written as a TODO.

I'll try to describe the prefetching mechanism as how I have implemented it right now:

- When a .db file is requested and downloaded, I save the link and the last time it got requested in the db (MirrorDB table). 
Db links are hard to generate otherwise.
- When a package file whose name is structured as _name-version-subversion-arch.pkg.tar.zst_ is requested, I save name, version (merging version-subversion into version), arch and the local relative file path in the db (Package table). I also keep track of when it got prefetched and when got downloaded.
- When the prefetching routine is being called, as first I delete "old" MirrorDB entries (old = if not downloaded after ttl_unaccessed_in_days days, it gets deleted from the table) and dead packages. A package is assumed dead if:
    - after ttl_unaccessed_in_days days it had been updated, it hasn't been actively requested by a client
    - after ttl_unupdated_in_days days in which it had never been requested nor updated. Right now the default value is quite high, set at 300 days
- Then it starts to download the db files stored in the MirrorDB table into "tmp-db" folder, it extract them, parse them and once parsed, stored their relevant entries (package name, arch and version) in RepoPackage table. Then I delete all the "tmp-db" files.
- Then I do an inner join between Package table and RepoPackage table to find only the installed packages which have a different version (in the join condition, i check package name and arch to be equal, and version to be different). Once those have been found, the old packages files (with their signature files) gets deleted and the new version gets prefetched
- The prefetching function is the same handleRequest function adapted to use a single writer (the file one), and it calls a different function to store package data in the db (cause it would have to store different data)
- Packages are stored by using as primary key both the package name and the architecture at the moment.

Please, provide me any hints on things to improve/change/adapt to this prefetching mechanism, maybe there is something I am missing. Some questions I have 

- I don't know if the package name and the architecture are enough to identify a package uniquely
- Are there are different package naming conventions that would fail the regex?

Sorry for any english mistakes, I am quite tired now :smile:

**Comment by Focshole** [comment]
Ok, good news, I have finished writing the code and it seems to work, now tests are fine, test coverage is around 81%. There are no breaking changes with @anatol version, with a @anatol version's config file, it would run exactly as before, the prefetching code would not be executed at all.

Prefetching should work fine, I am testing pacoloco at home right now with default repositories, I'll provide a few updates in the following days (I plan to prefetch every night for a week to check if it does a good work or not).

I have added a sqlite3 database to keep track of the downloaded databases links, integrated with GORM. Couldn't keep all the relevant data in memory cause it could have been a bit overkill as memory consumption and even harder to handle. The database could be useful for deploying the "stats" thing written as a TODO.

I'll try to describe the prefetching mechanism as how I have implemented it right now:

-  When a .db file is requested and downloaded, I save the link and the last time it got requested in the db (MirrorDB table),  db links are hard to generate otherwise.
- When a package file whose name is structured as name-version-subversion-arch.pkg.tar.zst is requested, I save name, version (merging version-subversion into version), arch and the local repo name in the db (Package table). I also keep track of when it got prefetched and when got downloaded.
- When the prefetching routine is being called, as first I delete "old" MirrorDB entries (old = if not downloaded after ttl_unaccessed_in_days days, it gets deleted from the table) and dead packages. A package is assumed dead if:
    - after ttl_unaccessed_in_days days it had been updated, it hasn't been actively requested by a client
    -  after ttl_unupdated_in_days days in which it had never been requested nor updated. Right now the default value is quite high, set at 300 days
- Then it starts to download the db files stored in the MirrorDB table into "tmp-db" folder in the pacoloco cache directory, it extract them, parse them and once parsed, stored their relevant entries (package name, arch and version) in RepoPackage table. Then I delete all the "tmp-db" files.
- Then I do an inner join between Package table and RepoPackage table to find only the installed packages which have a different version upstream (in the join condition, i check package name and arch to be equal, and version to be different). Once those have been found, the old packages files (with their signature files) gets deleted and the new version gets prefetched
- The prefetching function is the same handleRequest function adapted to use a single writer (the file one), and it calls a different function to store package data in the db (cause it would have to store different data)
-  Packages are stored by using as primary key the package name, the architecture and the repository name.

Please, provide me any hints on things to improve/change/adapt to this prefetching mechanism, maybe there is something I am missing. Some questions I have
    Are there are different package naming conventions that would fail the regex?

**Comment by anatol** [comment]
The idea described in your previous comment sounds good. Looking forward to see a pull request with this feature.

Pulling gorm/sqlite3 dependency is fine. It makes storing/queuering package info simpler.

> different package naming conventions that would fail the regex?

One thing that comes to my mind is packages with `epoch` like for this package https://archlinux.org/packages/community/x86_64/gitlab-shell/ it should be handled by pacoloco.

> now tests are fine, test coverage is around 81%

Nice. Add as much tests as you can as it will pay back in the future.

**Comment by anatol** [comment]
> edit: .xz files still exist, adding support for them soon

There is quite a number of supported archive formats. See https://archlinux.org/pacman/makepkg.conf.5.html for the list of currently supported formats.
